pkg/deviceclaimingserver/enddevices/ttjs/ttjs.go

```diff
@@ -137,15 +164,24 @@ func (client *TTJS) Claim(ctx context.Context, req *ttnpb.ClaimEndDeviceRequest)
 		devEUI = authenticatedIDs.DevEui
 		ownerToken = authenticatedIDs.AuthenticationCode
 	} else if qrCode := req.GetQrCode(); qrCode != nil {
-		data, err := qrcode.Parse(qrCode)
+		qrg, err := client.getQRGeneratorClient(ctx)
+		if err != nil {
+			return nil, err
+		}
+		data, err := qrg.Parse(ctx, &ttnpb.ParseEndDeviceQRCodeRequest{
+			QrCode: qrCode,
+		})
 		if err != nil {
 			return nil, errParseQRCode.WithCause(err)
 		}
-		authIDs, ok := data.(qrcode.AuthenticatedEndDeviceIdentifiers)
-		if !ok {
-			return nil, errQRCodeData.New()
+
+		dev := data.GetEndDeviceTempate().GetEndDevice()
+		if dev == nil || dev.ClaimAuthenticationCode == nil {
+			return nil, errParseQRCode.New()
 		}
-		joinEUI, devEUI, ownerToken = authIDs.AuthenticatedEndDeviceIdentifiers()
+		joinEUI = *dev.GetIds().JoinEui
+		devEUI = *dev.GetIds().DevEui
+		ownerToken = dev.ClaimAuthenticationCode.Value
```

Comment:
I don't think that the The Things Join Server implementation should be bothered with parsing QR codes. This should be nothing more than a client to a Join Server.

If anything, DCS should do this before it let's JS clients claim.

In previous review rounds I already stressed on decoupling stuff. It would have been helpful if claimers are not taking any `pkg/ttnpb` types so that we are sure things aren't getting mixed up. You don't really need `ClaimEndDeviceRequest` here; JS claimers just need EUIs and a secret. It doesn't matter how does values were determined at this level.